src/devices/cellular.c

```diff
@@ -0,0 +1,534 @@
+#include "loggerConfig.h"
+#include "mod_string.h"
+#include "printk.h"
+#include "serial_buffer.h"
+#include "serial.h"
+#include "sara_u280.h"
+#include "sim900.h"
+#include "taskUtil.h"
+
+#include <stdbool.h>
+
+#define AUTH_MAX_JSMN_TOKENS	5
+
+static const struct cell_modem_methods* methods;
+static const struct at_config *at_cfg;
+
+static struct cellular_info cell_info;
+static struct telemetry_info telemetry_info;
+static const enum log_level serial_dbg_lvl = TRACE;
+
+static const struct at_config* get_safe_at_config()
+{
+        /* Safe but un-optimal values for all modems */
+        static const struct at_config cfg = {
+                .urc_delay_ms = 250,
+        };
+
+        return &cfg;
+}
+
+static void cell_serial_tx_cb(const char *data)
+{
+        static bool pr_tx_pfx = true;
+
+        for(; *data; ++data) {
+                if (pr_tx_pfx) {
+                        printk(serial_dbg_lvl, "[cell] tx: ");
+                        LED_toggle(0);
+                        pr_tx_pfx = false;
+                }
+
+                switch(*data) {
+                case('\r'):
+                case('\n'):
+                        printk_crlf(serial_dbg_lvl);
+                        pr_tx_pfx = true;
+                        break;
+                default:
+                        printk_char(serial_dbg_lvl, *data);
+                        break;
+                }
+        }
+}
+
+static void cell_serial_rx_cb(const char *data)
+{
+        static bool new_line = true;
+
+        if (NULL == data)
+                return;
+
+        for(; *data; ++data) {
+                if (new_line) {
+                        printk(serial_dbg_lvl, "[cell] rx: ");
+                        new_line = false;
+                }
+
+                switch(*data) {
+                case('\r'):
+                        printk(serial_dbg_lvl, "<cr>");
+                        break;
+                case('\n'):
+                        printk(serial_dbg_lvl, "<lf>\r\n");
+                        new_line = true;
+                        break;
+                default:
+                        printk_char(serial_dbg_lvl, *data);
+                        break;
+                }
+        }
+}
+
+size_t cellular_exec_cmd(struct serial_buffer *sb,
+                         const size_t wait,
+                         const char **rx_msgs,
+                         const size_t rx_msgs_size)
+{
+        serial_buffer_append(sb, "\r");
+        serial_buffer_tx(sb);
+        serial_buffer_clear(sb);
+        const size_t rv = serial_buffer_rx_msgs(sb, wait, rx_msgs,
+                                                rx_msgs_size);
+
+        /*
+         * URC Pause (must be > 20ms)
+         * TODO: Add support for handling URCs.  Right now we ignore them.
+         *       This will likely require a re-write of how cellular is
+         *       done.
+         */
+        if (NULL == at_cfg)
+                at_cfg = get_safe_at_config();
```

Comment:
Currently this only sets the backoff time on AT commands.  Ublox only needs 25ms where as sim900 need 250 (because it is slooooooowwwwww).